reyna/src/com/b2msolutions/reyna/Dispatcher.java

```diff
@@ -52,17 +61,32 @@ protected Result sendMessage(Message message, HttpPost httpPost, HttpClient http
     public static Result canSend(Context context) {
         ConnectivityManager connectivityManager = (ConnectivityManager) context.getSystemService(Context.CONNECTIVITY_SERVICE);
         NetworkInfo info = connectivityManager.getActiveNetworkInfo();
-        if(info == null || !info.isConnectedOrConnecting()) {
+        if (info == null || !info.isConnectedOrConnecting()) {
             return Result.NOTCONNECTED;
         }
+        int type = info.getType();
```

Comment:
I would like to suggest a refactoring here.
Can we have an interface or class called Blackout that expose one method called canSend(Context). then we have multiple implementation for this interface, so we hace WlanBlackout, WWan Blackout, PoweBlackout, etc...

in the main Dispatcher  canSend. we will initiate an array of Balckout objects and just making an And operation.

I think this will make the code more clear as we don't need to have methods that shows all combinations. like cantSendOnRoamingCharging, cantSendOnWlanCharging

there is a missing cases when device is conencted to both WLAN and WWAN then power.

interface IBlackout{
boolean canSend(Context context)
}

class WlanBlackout implments IBlackout{
boolean canSend(Context context) {
 if (type == ConnectivityManager.TYPE_WIFI) {
            if(!blackoutTime.canSendOnWlan(new GregorianCalendar())) {
                return true;
            }
        }
}
}


class PowerBlackout implments IBlackout{
boolean canSend(Context context) {
 if (power.isCharging(context)) {
            if(!preferences.canSendOnCharge()) {
                return true;
            }
        }
}
}


Dispatcher
  public static Result canSend(Context context) {
	  IBlackout[] Blackouts= new IBlackout[5] {new WlanBlackout, newPowerBlackout}
	  foreach(IBlackout blackout : Blackouts)
	  {
		  if (!blackout.canSend)
		  {
			  return BLACKOUT
		  }
	  }
  }